controllers/clusters/cassandra_controller.go

```diff
@@ -400,30 +316,122 @@ func (r *CassandraReconciler) handleUpdateCluster(
 		return reconcile.Result{}, err
 	}
 
-	iCassandra, err := c.FromInstAPI(iData)
+	iCassandra := v1beta1.Cassandra{}
+	iCassandra.FromInstAPI(instaModel)
+
+	bootstrap := newOnPremisesBootstrap(
+		r.Client,
+		c,
+		r.EventRecorder,
+		iCassandra.Status.ToOnPremises(),
+		c.Spec.OnPremisesSpec,
+		newExposePorts(c.GetExposePorts()),
+		c.GetHeadlessPorts(),
+		c.Spec.PrivateNetwork,
+	)
+
+	ctx := context.Background()
+
+	err = handleCreateOnPremisesClusterResources(ctx, bootstrap)
 	if err != nil {
 		l.Error(
-			err, "Cannot convert cluster from the Instaclustr API",
+			err, "Cannot create resources for on-premises cluster",
+			"cluster spec", c.Spec.OnPremisesSpec,
+		)
+		r.EventRecorder.Eventf(
+			c, models.Warning, models.CreationFailed,
+			"Resources creation for on-premises cluster is failed. Reason: %v",
+			err,
+		)
+		return reconcile.Result{}, err
+	}
+
+	err = r.startClusterOnPremisesIPsJob(c, bootstrap)
+	if err != nil {
+		l.Error(err, "Cannot start on-premises cluster IPs check job",
+			"cluster ID", c.Status.ID,
+		)
+
+		r.EventRecorder.Eventf(
+			c, models.Warning, models.CreationFailed,
+			"On-premises cluster IPs check job is failed. Reason: %v",
+			err,
+		)
+		return reconcile.Result{}, err
+	}
+
+	l.Info(
+		"On-premises resources have been created",
+		"cluster name", c.Spec.Name,
+		"on-premises Spec", c.Spec.OnPremisesSpec,
+		"cluster ID", c.Status.ID,
+	)
+
+	return models.ExitReconcile, nil
+}
+
+func (r *CassandraReconciler) handleCreateCluster(
+	ctx context.Context,
+	l logr.Logger,
+	c *v1beta1.Cassandra,
+) (reconcile.Result, error) {
+	l = l.WithName("Cassandra creation event")
+	if c.Status.ID == "" {
+		err := r.createCluster(ctx, c, l)
+		if err != nil {
+			r.EventRecorder.Eventf(c, models.Warning, models.CreationFailed, "Failed to create cluster. Reason: %v", err)
+			return reconcile.Result{}, err
+		}
+	}
+
+	if c.Status.State != models.DeletedStatus {
+		err := r.startClusterJobs(c, l)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to start cluster jobs, err: %w", err)
+		}
+	}
+
+	if c.Spec.OnPremisesSpec != nil && c.Spec.OnPremisesSpec.EnableAutomation {
+		return r.handleOnPremises(c, l)
+	}
+
+	return models.ExitReconcile, nil
+}
```

Comment:
So if cluster status is deleted. With your logic it will just continue executing, right?

Reply:
When I started refactoring I decided to try to keep flow as it has been before. Anyway I agree it looks a bit strange and there is cassandra details api call inside `handleOnPremises` method.
I decided to move it inside the if statement above.